earthaccess/search.py

```diff
@@ -11,6 +11,29 @@
 from .results import DataCollection, DataGranule
 
 
+def _normalize_datetime(raw: None | str | datetime) -> None | datetime:
+    # empty string or None is None
+    if not raw:
+        return None
+    # the cmr.*Query.temporal method only parses a subset of ISO 8601 strings, so we
+    # add flexibility, inclusive of the presence or absence of timezone information
+    try:
+        dt = dateutil.parser.parse(raw)
```

Comment:
I agree with @mfisher87 that `isinstance` is a good check here, but I think you can simplify a bit by providing a default to `dateutil.parser.parse`. 

```python
def _normalize_datetime(dt: None | str | datetime) -> None | datetime:
    # empty string or None is None
    if not raw:
        return None
    
    # the cmr.*Query.temporal method only parses a subset of ISO 8601 strings, so we
    # add flexibility, inclusive of the presence or absence of timezone information, defaulting to UTC
    if not isinstance(dt, datetime):
        dt = dateutil.parser.parse(dt, default=datetime(1, 1, 1, 0, 0, 0, 0, timezone.utc))

    # and then we guarantee a UTC datetime
    return dt.astimezone(timezone.utc)
```

Notably, providing a default will also round down for partial datetimes, which from a search perspective, is what we want I think. Examples of providing a default:

```python
>>> from datetime import datetime, timezone
>>> from dateutil.parser import parse as date_parser
>>> 
>>> date_parser('2024')
datetime.datetime(2024, 3, 1, 0, 0)
>>> 
>>> date_parser('2024', default=datetime(1, 1, 1, 0, 0, 0, 0, timezone.utc))
datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)
>>> 
>>> date_parser('20240405')
datetime.datetime(2024, 4, 5, 0, 0)
>>> 
>>> date_parser('20240405', default=datetime(1, 1, 1, 0, 0, 0, 0, timezone.utc))
datetime.datetime(2024, 4, 5, 0, 0, tzinfo=datetime.timezone.utc)
```

It might also be worth handling `date` -> `datetime`.

Reply:
Though rounding down assumes the date is used for "search start"; you'd want to round up for "search stop"

Reply:
>It might also be worth handling `date` -> `datetime`.

:100:

I love using plain `date` objects!

Reply:
Thanks for the help! Now handles `date` objects. Leaving as draft until I take another look (running to meeting!).

Frankly, I don't grok typing, but following your lead. The `try...except` seems more Pythonic (duck typing) than having to annotate with every class that might implement `strftime`. I'll keep learning.

@jhkennedy I have to disagree on your simplification and on rounding up. On the simplification, we can't use `astimezone` on naive datetime objects if we want to assume UTC rather than assume local. We could argue about that, but I'd push heavily towards assuming people searching earthdata assume UTC. On rounding, which now happens either at parse time or when combining a `date` with `dt.time()`, both go down. This is quite expected, and I would be confused if my "2024-03" got interpreted as "2024-03-31T23:59:59" .

Reply:
>The `try...except` seems more Pythonic (duck typing)

I would have agreed with that in the pre-typechecking era of Python wholeheartedly. The "easier to ask forgiveness than permission" approach, i.e. just try to do stuff and deal with the fallout if there is any, was a Python style tenet that I really struggled to adopt as an early learner. I was used to writing exhaustive checks to protect my code that knows exactly what kind of objects it's dealing with. I think that familiarity is part of why I've found it so fun and natural to move to a typed Python style. It does kind of feel like it contradicts "Pythonic", but at the same time the rate of adoption of type annotations in the Python ecosystem has been _phenominal_. My way of thinking is that "Pythonic" is a consensus thing, not a set-in-stone thing. It just means something slightly different these days.

Reply:
@itcarroll Yes, in my simplification, I missed `datetime` inputs with no time zone -- we do want to assume UTC if not otherwise specified.

 On rounding, the current implementation in the PR *does not* round down -- you need to provide a default `datetime` if you want `dateutil.parser.parse` to round down. If no month or day is provided, they will default to the current month/day:
 ```python
 >>> import datetime as dt
>>> import dateutil.parser
>>> 
>>> datetutil.parser.parse('2024')
datetime.datetime(2024, 3, 1, 0, 0)
>>> 
>>> datetutil.parser.parse('2024', default=datetime(1, 1, 1, 0, 0, 0, 0, dt.timezone.utc))
datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)
```

On rounding up, I was specifically talking about the search stop time in [temporal range searches](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#temporal-range-searches). That is, when providing a range like `['2023', '2024']`
* rounding the stop date down will only include granules in 2023 (exclusive bound)
* *rounding the stop date up would return granules in 2023 and 2024 (inclusive bound). 

Now that I think about it more, I agree rounding down is always preferable and in-line with python ranges and CMR expectations

I also think this function is getting complicated enough that it deserves a set of tests itself -- do you want me to push tests for it specifically?

Reply:
On rounding, the current implementation in the PR does not round down

Oh, sorry @jhkennedy, I missed that!

Yes, like anything having to do [with Python datetimes](https://stackoverflow.com/questions/13703720/converting-between-datetime-timestamp-and-datetime64#comment67298380_21916253), this is getting complicated. Please feel free to add any necessary tests.

Reply:
:thinking: more on rounding -- I noticed #330 actually asks for *round up* for the stop date / `Date_To` parameters and implies this is the behavior for Earthdata Search. I haven't tested the CMR or Earthdata Search behavior; do you know what it does here?

Reply:
I still think rounding up will be confusing, see question to OP on 190. I'm not going to try to implement the recurring/periodic search in 488, but hopefully we can close 190 one way or the other.

Reply:
I agree, and that was a fantastic write up on #190. Down is also more Pythonic.

Reply:
Default provided that rounds down. Will reconsider if a reply comes to #190, so leaving as draft for now.